File: Data Structures & Algorithms/longest-repeating-substring-with-replacement/submission-0.py
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        left = 0
        longest = 0
        freq = collections.defaultdict(int)
        most_frequent = 0

        for right in range(len(s)):
            freq[s[right]] += 1
            most_frequent = max(most_frequent, freq[s[right]])

            while ((right - left + 1) - most_frequent) > k:
                freq[s[left]] -= 1
                left += 1


            longest = max(longest, right - left + 1)
        
        return longest

File: Data Structures & Algorithms/reverse-words-in-a-string-ii/submission-0.go
// ["t","h","e"," ","s","k","y"," ","i","s"," ","b","l","u","e"]
// ["e","u","l","b"," ","s","i"," ","y","k","s"," ","e","h","t"]
// ["b","l","u","e"," ","s","i"," ","y","k","s"," ","e","h","t"]

func reverseWords(s []byte) {
	left, right := 0, len(s) - 1

	for left < right {
		s[left], s[right] = s[right], s[left]
		left++
		right--
	}

	l := 0
	for l < len(s) {
		r := l
		for r < len(s) && s[r] != ' ' {
			r++
		}

		wl, wr := l, r-1
		for wl < wr {
			s[wl], s[wr] = s[wr], s[wl]
			wl++
			wr--
		}

		l = r+1
	}
}
